boden/grund.cc

```diff
@@ -1883,94 +1883,88 @@ bool grund_t::weg_erweitern(waytype_t wegtyp, ribi_t::ribi ribi)
 }
 
+sint64 grund_t::remove_trees() {
+	return visit_trees_and_groundobjs(*this, tree_groundobj_remove_visitor);
+}
 
+// Land is for sale if all stationary objects on ground are
+// - not owned by another player (including public player)
+// - not public right of way
+bool grund_t::is_for_sale() const {
+	const obj_t* const obj0 = obj_bei(0);
+	const weg_t* const way1 = get_weg_nr(1);
+	return (!obj0 || obj0->is_moving() || (obj0->get_owner()==nullptr && (obj0->get_typ() != obj_t::way || !((const weg_t*)obj0)->is_public_right_of_way()))) &&	// First obj, if exists, is a stationary object not owned by a player nor prow
+		   (!way1 || !way1->is_public_right_of_way());	// Second object, if exists is
+}
 
-sint64 grund_t::neuen_weg_bauen(weg_t *weg, ribi_t::ribi ribi, player_t *player, koord3d_vector_t *route)
+//For now, ownership is defined as the owner of the first object, if that object is stationary except if that object is public right of way, then nobody owns that land.
+//The usual waytype order will determine ownership, not the order of construction.
+//TODO: Store ownership, so the player who first claimed a land will own it until he sells his last object on that tile.
+player_t* grund_t::get_owner() const {
+	const obj_t* const obj = obj_bei(0);
+	if(obj && !obj->is_moving() && (obj->get_typ() != obj_t::way || !((const weg_t*)obj)->is_public_right_of_way())){
+		return obj->get_owner();
+	}
+	return nullptr;
+}
+
+
+sint64 grund_t::neuen_weg_bauen(weg_t *weg, ribi_t::ribi ribi, player_t *player)
 {
 	sint64 cost=0;
 
 	// not already there?
 	const weg_t * alter_weg = get_weg(weg->get_waytype());
-	if(alter_weg==NULL) {
+	if(alter_weg==nullptr) {
 		// ok, we are unique
 		// Calculate the forge cost
-		sint64 forge_cost = welt->get_settings().get_forge_cost(weg->get_waytype());
+		cost -= welt->get_forge_cost(weg->get_waytype(), pos);;
 
-		if(route != NULL)
-		{
-			// No parallel way discounting for bridges and tunnels.
-			for(int n = 0; n < 8; n ++)
-			{
-				const koord kn = pos.get_2d().neighbours[n] + pos.get_2d();
-				if(!welt->is_within_grid_limits(kn))
-				{
-					continue;
-				}
-				const koord3d kn3d(kn, welt->lookup_hgt(kn));
-				grund_t* to = welt->lookup(kn3d);
-				const weg_t* connecting_way = to ? to->get_weg(weg->get_waytype()) : NULL;
-				const ribi_t::ribi connecting_ribi = connecting_way ? connecting_way->get_ribi() : ribi_t::all;
-				if(route->is_contained(kn3d) || (ribi_t::is_single(ribi) && ribi_t::is_single(connecting_ribi)))
-				{
-					continue;
-				}
-				const grund_t* gr_neighbour = welt->lookup_kartenboden(kn);
-				if(gr_neighbour && gr_neighbour->get_weg(weg->get_desc()->get_waytype()))
-				{
-					// This is a parallel way of the same type - reduce the forge cost.
-					forge_cost *= welt->get_settings().get_parallel_ways_forge_cost_percentage(weg->get_waytype());
-					forge_cost /= 100ll;
-					break;
-				}
-			}
-		}
-		cost -= forge_cost;
```

Comment:
Replace inline forge cost calculation by a call to get_forge_cost (remove duplicate code)